### ISSUE TITLE
semver-checks fails to detect semver change in core-foundation-sys

### ISSUE THREAD

**Issue by jrmuizel** [obi1kenobi/cargo-semver-checks#473]
### Steps to reproduce the bug with the above code

Run `cargo semver-checks check-release` in the `core-foundation-sys` directory of https://github.com/servo/core-foundation-rs @ eb00adae0c6d3694dccc54b195132c2929ac8c96 

### Actual Behaviour

No changes detected

### Expected Behaviour

semver-checks fails to detect that `UniChar` has moved from `string::` to `base::`

#### Software version

cargo-semver-checks 0.21.0

#### Operating system

Mac OS X 10.15.6 (Darwin 19.6.0)

#### Command-line

```bash
/Users/jrmuizel/.cargo/bin/cargo-semver-checks semver-checks --bugreport
```

#### cargo version

```
> cargo -V
cargo 1.70.0 (ec8a8a0ca 2023-04-25)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-apple-darwin
- Family: unix
- OS: macos
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2,sse3,ssse3
- Host: x86_64-apple-darwin

**Comment by jrmuizel** [comment]
Actually, semver-checks will detect
```
  --- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-check/tree/v0.21.0/src/lints/function_missing.ron

Failed in:
  function core_foundation_sys::url::CFURLStopAccessingSecurityScopedResource, previously in file /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/core-foundation-sys-0.8.4/src/url.rs:162
  function core_foundation_sys::url::CFURLStartAccessingSecurityScopedResource, previously in file /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/core-foundation-sys-0.8.4/src/url.rs:161
  ```
  
  but not the moved type mentioned above.

**Comment by obi1kenobi** [comment]
Thanks for the report.

This is a known issue, caused in part by the same limitation that also causes #355: `cargo-semver-checks` today cannot see or analyze types defined outside the crate being scanned. `UniChar` is defined as:
```rust
use std::os::raw::{c_uint, c_void, c_int, c_short, c_uchar, c_ushort};
// ...
pub type UniChar = c_ushort;
```
and `c_ushort` is an external type.

The limitation is due to the fact that external types' definitions are not part of the rustdoc JSON file of the scanned crate. There is currently no good way to connect data in multiple rustdoc JSON files together.

The rustdoc team is aware of this limitation and is interested in resolving it; I've stayed in touch with them on this.

Sorry I don't have a better answer!

**Comment by jrmuizel** [comment]
Is there an issue tracking the rustdoc problem?

**Comment by obi1kenobi** [comment]
This is the closest thing to a single tracking issue I could find: https://github.com/rust-lang/compiler-team/issues/635

That's a prerequisite `rustc` change after which rustdoc JSON will probably tweak the format in some way that resolves the problems in [this link](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Identifying.20external.20crates.20in.20Rustdoc.20JSON/near/352551996), which is also mentioned in the issue.

**Comment by jrmuizel** [comment]
We ended up getting burnt with this with the release of core-foundation-sys 0.8.5. https://github.com/servo/core-foundation-rs/issues/619

a `pub use CFIndex` was remove in https://github.com/servo/core-foundation-rs/commit/6d00383bafd95f7457a288e1948fc4111a34ae9e which broke semver and was not detected.

**Comment by jrmuizel** [comment]
@obi1kenobi the mentioned compiler-team issue has been closed. What's next?

**Comment by obi1kenobi** [comment]
Unfortunately that just means the compiler MCP is complete. AFAIK none of the work along that MCP has started yet, and I don't think there's a tracking issue for it either.

I don't think anyone is funded to do that work at the moment, so it's dependent on volunteers allocating time to it. Unless there's funding to be found, I don't think there's a good way to nudge it forward.
